presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotFilterExpressionConverter.java

```diff
@@ -89,15 +100,50 @@ private PinotExpression handleLogicalBinary(
         }
         List<RowExpression> arguments = call.getArguments();
         if (arguments.size() == 2) {
-            return derived(format(
-                    "(%s %s %s)",
-                    arguments.get(0).accept(this, context).getDefinition(),
-                    operator,
-                    arguments.get(1).accept(this, context).getDefinition()));
+            String def0 = arguments.get(0).accept(this, context).getDefinition();
+            String def1 = arguments.get(1).accept(this, context).getDefinition();
+            def1 = handleTimeStampAndDateCast(arguments.get(0), def1);
+            def0 = handleTimeStampAndDateCast(arguments.get(1), def0);
+            return derived(format("(%s %s %s)", def0, operator, def1));
         }
         throw new PinotException(PINOT_UNSUPPORTED_EXPRESSION, Optional.empty(), format("Unknown logical binary: '%s'", call));
     }
 
+    private String handleTimeStampAndDateCast(RowExpression rowExpression, String definition)
+    {
+        if (!(rowExpression instanceof CallExpression)) {
+            return definition;
+        }
+        CallExpression callExpr = (CallExpression) rowExpression;
+        if (!standardFunctionResolution.isCastFunction(callExpr.getFunctionHandle())) {
+            return definition;
+        }
+        if (callExpr.getArguments().size() != 1) {
+            return definition;
+        }
+        Type inputType = callExpr.getArguments().get(0).getType();
+        Type expectedType = callExpr.getType();
+        if (inputType == DateType.DATE && (expectedType == TimestampType.TIMESTAMP || expectedType == TimestampWithTimeZoneType.TIMESTAMP_WITH_TIME_ZONE)) {
+            try {
+                Long daysSinceEpoch = TimeUnit.MILLISECONDS.toDays(Long.parseLong(definition));
```

Comment:
`definition` is actually a PQL expression fragment corresponding to a whole expression subtree, and need not be a number. IMHO, Its a bit brittle to first cast an expression tree to a PQL expression-string and then assume that is parseable as a number.

Reply:
You are right, this logic only handles the literal pushdown, otherwise we will fallback to row retrieval and relies on presto to do the filtering.